### PR TITLE
cql3: expr: detemplate and deinline find_in_expression() and count_if()

### PR DESCRIPTION
**Pull request by avikivity**
The expression code provides some utilities to examine and manipulate
expressions at prepare time. These are not (or should not be) in the fast
path and so should be optimized for compile time and code footprint
rather than run time.

This series does so by detemplating and deinlining find_in_expression()
and count_if().

### CODE REVIEW

**Comment by avikivity** [comment]
Note: find_in_expression() is mistakenly still used in a fast path, but @cvybhu [promised](https://github.com/scylladb/scylla/pull/9481#issuecomment-982698208) to fix that.

**Comment by cvybhu** [comment]
Why do we want to deinline those? Shouldn't the compiler be able to decide when to inline stuff?

**Comment by avikivity** [comment]
> Why do we want to deinline those? Shouldn't the compiler be able to decide when to inline stuff?

So the compiler doesn't have to see them in every translation unit, and recompile any time it changes. These are prepare path operations and not critical.

**Review by cvybhu — CHANGES_REQUESTED** [review]
Looks ok, although I'm not sure if complicating the code to get a bit quicker compilation is worth it.

The number of instructions increased a bit, but that might go away once `need_filtering()` is moved to prepare time:
```bash
median 34179.54 tps ( 75.1 allocs/op,  12.1 tasks/op,   45944 insns/op)
median 34274.65 tps ( 75.1 allocs/op,  12.1 tasks/op,   46424 insns/op)
# build/release/test/perf/perf_simple_query --cpuset 1 -m 1G --task-quota-ms 10  --random-seed 0 --operations-per-shard 1000000
```

File: cql3/expr/expression.cc [hunk 1]
```diff
@@ -1093,6 +1093,78 @@ expression replace_token(const expression& expr, const column_definition* new_cd
     });
 }
 
+bool recurse_until(const expression& e, const noncopyable_function<bool (const expression&)>& predicate_fun) {
+    if (auto res = predicate_fun(e)) {
+        return res;
+    }
+    return expr::visit(overloaded_functor{
+            [&] (const binary_operator& op) {
+                if (auto found = recurse_until(op.lhs, predicate_fun)) {
```
[the remark this answers is not in the document]
Comment (avikivity) [reply cql3/expr/expression.cc:1102]: Yes, I kept the changes minimal.

[the remark this answers is not in the document]
Comment (avikivity) [reply cql3/expr/expression.cc:1102]: (you're reviewing a code movement patch, not the code changing patch)

**Review by psarna — APPROVED** [review]
Looks good, and btw - @cvybhu compilation times aside, the other important effect of such changes is reduced compiled code size, which also influences performance

**Pull request closed without merging**